lib/ansible/galaxy/role.py

```diff
@@ -412,24 +452,42 @@ def install(self):
                                         # to debug a broken installation.
                                         if not n_part:
                                             continue
-                                        if n_part == '..':
+                                        if n_part == '..' and attr == 'name':  # we validate linkname above, links must not resolve to a path outside the role
+                                            invalid_parts |= True
                                             display.warning(f"Illegal filename '{n_part}': '..' is not allowed")
                                             continue
                                         if n_part.startswith('~'):
+                                            invalid_parts |= True
                                             display.warning(f"Illegal filename '{n_part}': names cannot start with '~'")
                                             continue
                                         if '$' in n_part:
+                                            invalid_parts |= True
                                             display.warning(f"Illegal filename '{n_part}': names cannot contain '$'")
                                             continue
                                         n_final_parts.append(n_part)
+
+                                    if invalid_parts or not n_final_parts:
+                                        raise AnsibleError(f"role content {member.name} is not able to be extracted")
```

Comment:
This is a small change to try to avoid installing broken roles, but is not backwards compatible. Should this affect `name` and `linkname`?